### PR TITLE
nixos/paperless: add 'database.createLocally'

### PR DESCRIPTION
**Pull request by ambroisie**
Make the module declare its dependency on `postgresql.service` when configured to use a local PostreSQL DB.

This should _probably_ live behind an option? Should we go full `setupPostgresqlDB`, to also configure the DB?

Uploading as a WIP, as I (naively) added the dependency to _all_ services in the module, and would like the maintainers to weigh in with their opinion before going further.

EDIT: added `database.createLocally` to configure and provision the PostgreSQL DB.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
> This should _probably_ live behind an option? Should we go full `setupPostgresqlDB`, to also configure the DB?

Other services name it like `services.mastodon.database.createLocally` or `services.miniflux.createDatabaseLocally`. Please use something similar

File: nixos/modules/services/misc/paperless.nix [hunk 1]
```diff
@@ -320,13 +323,23 @@ in
           echo "$superuserState" > "$superuserStateFile"
         fi
       '';
-    } // optionalAttrs enableRedis {
-      after = [ "redis-paperless.service" ];
+      requires = [ ]
+        ++ lib.optional localPostgres "postgresql.service"
+        ;
+      after = [ ]
+        ++ lib.optional enableRedis "redis-paperless.service"
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/paperless.nix:330]:
```suggestion
      requires = lib.optional localPostgres "postgresql.service";
      after = lib.optional enableRedis "redis-paperless.service"
```

File: nixos/modules/services/misc/paperless.nix [hunk 2]
```diff
@@ -361,7 +379,12 @@ in
       # Bind to `paperless-scheduler` so that the web server never runs
       # during migrations
       bindsTo = [ "paperless-scheduler.service" ];
-      after = [ "paperless-scheduler.service" ];
+      requires = [ ]
+        ++ lib.optional localPostgres "postgresql.service"
+        ;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/paperless.nix:384]:
```suggestion
      requires = lib.optional localPostgres "postgresql.service";
```

File: nixos/modules/services/misc/paperless.nix [hunk 3]
```diff
@@ -344,7 +357,12 @@ in
       # Bind to `paperless-scheduler` so that the consumer never runs
       # during migrations
       bindsTo = [ "paperless-scheduler.service" ];
-      after = [ "paperless-scheduler.service" ];
+      requires = [ ]
+        ++ lib.optional localPostgres "postgresql.service"
+        ;
+      after = [ "paperless-scheduler.service" ]
+        ++ lib.optional localPostgres "postgresql.service"
+        ;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/paperless.nix:365]:
```suggestion
      requires = lib.optional localPostgres "postgresql.service";
      after = [ "paperless-scheduler.service" ]
        ++ lib.optional localPostgres "postgresql.service";
```

File: nixos/modules/services/misc/paperless.nix [hunk 4]
```diff
@@ -320,13 +323,23 @@ in
...
     };
 
     systemd.services.paperless-task-queue = {
       description = "Paperless Celery Workers";
-      after = [ "paperless-scheduler.service" ];
+      requires = [ ]
+        ++ lib.optional localPostgres "postgresql.service"
+        ;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/paperless.nix:339]:
```suggestion
      requires = lib.optional localPostgres "postgresql.service";
```

**Review by SuperSandro2000 — COMMENTED** [review]
otherwise good idea

**Comment by ambroisie** [comment]
> Other services name it like services.mastodon.database.createLocally or services.miniflux.createDatabaseLocally. Please use something similar

Sure, I wasn't sure which name was most popular, just used the first one I found on NixOS search :-).

**Comment by ambroisie** [comment]
Haven't been able to run the NixOS test on my current computer, which doesn't have KVM.

Will rely on CI for testing.

File: nixos/modules/services/misc/paperless.nix [hunk 5]
```diff
@@ -248,11 +247,34 @@ in
...
+      enable = true;
+      ensureDatabases = [ "paperless" ];
+      ensureUsers = [{
+        name = config.services.paperless.user;
+        ensureDBOwnership = true;
+      }];
+    };
+
+    services.paperless.settings = lib.mkIf cfg.database.createLocally {
+      PAPERLESS_DBHOST = "/run/postgresql";
```
Comment (ambroisie) [inline nixos/modules/services/misc/paperless.nix:275]: Perhaps should also include `PAPERLESS_DBNAME = "paperless"`?

Comment (ambroisie) [reply nixos/modules/services/misc/paperless.nix:275]: Added all of `PAPERLESS_DBENGINE`, `PAPERLESS_DBNAME`, `PAPERLESS_DBUSER` as well.

**Comment by ambroisie** [comment]
@SuperSandro2000 @leona-ya friendly ping :-).

**Review by leona-ya — APPROVED** [review]

**Pull request merged**